doc/sphinx/source/libCEEDapi.rst

```diff
@@ -0,0 +1,508 @@
+impact many applications from a single implementation.
+
+Our long-term vision is to include a variety of backend implementations in
+libCEED, ranging from reference kernels to highly optimized kernels targeting
+specific devices (e.g. GPUs) or specific polynomial orders. A simple reference
+backend implementation is provided in the file
+`ceed-ref.c <https://github.com/CEED/libCEED/blob/master/backends/ref/ceed-ref.c>`_.
+
+On the frontend, the mapping between the decomposition concepts and the code
+implementation is as follows:
+
+- **L-**, **E-** and **Q-vector** are represented as variables of type :ref:`CeedVector`.
+  (A backend may choose to operate incrementally without forming explicit **E-** or
+  **Q-vectors**.)
+
+- :math:`\mathbf{G}` is represented as variable of type :ref:`CeedElemRestriction`.
+
+- :math:`\mathbf{B}` is represented as variable of type :ref:`CeedBasis`.
+
+- the action of :math:`\mathbf{D}` is represented as variable of type :ref:`CeedQFunction`.
+
+- the overall operator :math:`\mathbf{G}^T \mathbf{B}^T \mathbf{D} \mathbf{B} \mathbf{G}`
+  is represented as variable of type
+  :ref:`CeedOperator` and its action is accessible through ``CeedOperatorApply()``.
+
+To clarify these concepts and illustrate how they are combined in the API,
+consider the implementation of the action of a simple 1D mass matrix
+(cf. `tests/t500-operator.c <https://github.com/CEED/libCEED/blob/master/tests/t500-operator.c>`_).
+
+.. literalinclude::  ../../../tests/t500-operator.c
+   :language: c
+   :linenos:
+
+The constructor
+
+.. code-block:: c
+
+   CeedInit(argv[1], &ceed);
+
+creates a logical device ``ceed`` on the specified *resource*, which could also be
+a coprocessor such as ``"/nvidia/0"``. There can be any number of such devices,
+including multiple logical devices driving the same resource (though performance
+may suffer in case of oversubscription). The resource is used to locate a
+suitable backend which will have discretion over the implementations of all
+objects created with this logical device.
+
+The ``setup`` routine above computes and stores :math:`\mathbf{D}`, in this case a
+scalar value in each quadrature point, while ``mass`` uses these saved values to perform
+the action of :math:`\mathbf{D}`. These functions are turned into the ``CeedQFunction``
+variables ``qf_setup`` and ``qf_mass`` in the ``CeedQFunctionCreateInterior()`` calls:
+
+.. code-block:: c
+
+   CeedQFunctionCreateInterior(ceed, 1, setup, setup_loc, &qf_setup);
+   CeedQFunctionAddInput(qf_setup, "_weight", 1, CEED_EVAL_WEIGHT);
+   CeedQFunctionAddInput(qf_setup, "dx", 1, CEED_EVAL_GRAD);
+   CeedQFunctionAddOutput(qf_setup, "rho", 1, CEED_EVAL_NONE);
+
+   CeedQFunctionCreateInterior(ceed, 1, mass, mass_loc, &qf_mass);
+   CeedQFunctionAddInput(qf_mass, "rho", 1, CEED_EVAL_NONE);
+   CeedQFunctionAddInput(qf_mass, "u", 1, CEED_EVAL_INTERP);
+   CeedQFunctionAddOutput(qf_mass, "v", 1, CEED_EVAL_INTERP);
```

Comment:
We've had issue is the past where these manually entered code snipits become decopuled from the tested code. In the past we've have periodically incorrect code in this file which is why we added the self-updating version.

Reply:
I think we want [literalinclude](https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-literalinclude), perhaps with `start-after`.

Reply:
I understand this is not the most desired option for maintanability and, in fact, it was not my first choice (see my [comment](https://github.com/CEED/libCEED/pull/446#issuecomment-581719177) above). Unfortunately the doxygen `\include` and `\snippet` doxygen directives are not recognized at all by any markdown parser for sphinx. So that seems not to be an option.

Reply:
Fixed with the suggested options. Thanks @jedbrown !

Reply:
Huzzah. I know I've failed to update docs before so I'm glad this works!